### ISSUE TITLE
Project Update fails with no output

### ISSUE THREAD

**Issue by hjkatz** [ansible/awx#9371]
##### ISSUE TYPE
 - Bug Report

##### ENVIRONMENT
* AWX version: 17.0.1
* AWX install method: kubernetes
* Ansible version:  2.9

##### STEPS TO REPRODUCE

Launch a Job Template with a custom SCM branch. Have that SCM branch been recently updated via a `git commit --amend` and `git push --force`, therefore forcing a rebase or a `git pull --force` to take place.

##### EXPECTED RESULTS

The project should update properly and the job should run.

##### ACTUAL RESULTS

No output in the Job screen: 
![image](https://user-images.githubusercontent.com/2234516/108566967-4e2b6780-72d5-11eb-925a-08a0b30a6ee7.png)

Some api information from `/api/v2/jobs/3340/`
```
    "job_explanation": "Previous Task Failed: {\"job_type\": \"project_update\", \"job_name\": \"Shared Ansible Repo\", \"job_id\": \"3341\"}",
    "execution_node": "awx-job-3340",
    "controller_node": "awx-65b9694b58-7chbw",
    "result_traceback": "Traceback (most recent call last):\n  File \"/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/main/tasks.py\", line 1385, in run\n    self.pre_run_hook(self.instance, private_data_dir)\n  File \"/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/main/tasks.py\", line 1944, in pre_run_hook\n    sync_task.run(local_project_sync.id)\n  File \"/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/main/tasks.py\", line 816, in _wrapped\n    return f(self, *args, **kwargs)\n  File \"/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/main/tasks.py\", line 1557, in run\n    raise AwxTaskError.TaskError(self.instance, rc)\nException: project_update 3341 (failed) encountered an error (rc=2), please see task stdout for details.\n",
    "event_processing_finished": true,
    "job_template": 35,
```

And from `/api/v2/project_updates/3341/`

```
    "job_args": "[\"bwrap\", \"--die-with-parent\", \"--unshare-pid\", \"--dev-bind\", \"/\", \"/\", \"--proc\", \"/proc\", \"--bind\", \"/tmp/ansible_runner_pi_utw897xe/ansible_runner_pi_ynztsn__/tmpbmfhfdbr\", \"/etc/ssh\", \"--bind\", \"/tmp/ansible_runner_pi_utw897xe/ansible_runner_pi_ynztsn__/tmp42e6stt5\", \"/etc/tower\", \"--bind\", \"/tmp/ansible_runner_pi_utw897xe/ansible_runner_pi_ynztsn__/tmpd4xxbyt8\", \"/tmp\", \"--bind\", \"/tmp/ansible_runner_pi_utw897xe/ansible_runner_pi_ynztsn__/tmp3_5_lb44\", \"/var/lib/awx\", \"--bind\", \"/tmp/ansible_runner_pi_utw897xe/ansible_runner_pi_ynztsn__/tmpnmjdmsbw\", \"/var/lib/awx/projects\", \"--bind\", \"/tmp/ansible_runner_pi_utw897xe/ansible_runner_pi_ynztsn__/tmpipv_v9sd\", \"/var/log\", \"--ro-bind\", \"/var/lib/awx/venv/ansible\", \"/var/lib/awx/venv/ansible\", \"--ro-bind\", \"/var/lib/awx/venv/awx\", \"/var/lib/awx/venv/awx\", \"--bind\", \"/tmp/awx_3341_7b88dg2g\", \"/tmp/awx_3341_7b88dg2g\", \"--bind\", \"/var/lib/awx/projects\", \"/var/lib/awx/projects\", \"--chdir\", \"/tmp/awx_3341_7b88dg2g/project\", \"ssh-agent\", \"sh\", \"-c\", \"ssh-add /tmp/awx_3341_7b88dg2g/artifacts/3341/ssh_key_data && rm -f /tmp/awx_3341_7b88dg2g/artifacts/3341/ssh_key_data && ansible-playbook -vvv -t update_git,install_roles,install_collections -i /tmp/awx_3341_7b88dg2g/inventory/hosts -e @/tmp/awx_3341_7b88dg2g/env/extravars project_update.yml\"]",
    "job_cwd": "/tmp/awx_3341_7b88dg2g/project",
 "job_explanation": "",
    "execution_node": "awx-65b9694b58-7chbw",
    "result_traceback": "",
    "event_processing_finished": true,
```

There is no `stdout` to see.

Here's the logs from `awx-task` from the AWX Pod:

```
2021-02-19 22:07:24,985 ERROR    awx.main.tasks job 3342 (running) Exception occurred while running task
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/main/tasks.py", line 1385, in run
    self.pre_run_hook(self.instance, private_data_dir)
  File "/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/main/tasks.py", line 1944, in pre_run_hook
    sync_task.run(local_project_sync.id)
  File "/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/main/tasks.py", line 816, in _wrapped
    return f(self, *args, **kwargs)
  File "/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/main/tasks.py", line 1557, in run
    raise AwxTaskError.TaskError(self.instance, rc)
Exception: project_update 3343 (failed) encountered an error (rc=2), please see task stdout for details.
```

Not sure what's going on here, this is all the information I can find.

**Comment by hjkatz** [comment]
I think this is happening when the playbook has a syntax error and nothing is reported.

When I run the playbook locally I see the following:

```
ansible-playbook -i inventory.stage playbooks/vmware/delete-vm.yml -e "hosts=[]"

...

ERROR! 'ldap_user' is undefined
```

And my playbook _does_ have a syntax error in `vars_prompt`

```
vars_prompt:
- name: ldap_user
  ...
- name: ldap_password
  prompt: "Enter the LDAP password of {{ ldap_user }}"
```

Since `ldap_user` does not yet exist for the `ldap_password` prompt, `ansible-playbook` throws an error.

Nonetheless, AWX should report this.

**Comment by unlikelyzero** [comment]
@hjkatz I think you're running into this issue: https://github.com/ansible/awx/issues/8788. Can you confirm?

**Comment by hjkatz** [comment]
No, if anything it's still related to this: https://github.com/ansible/awx/issues/8709

This output is still not shown in AWX 17.0.1

**Comment by jakemcdermott** [comment]
:wave: Some failed jobs just don't have any stdout events to show. Can you check the api to see that there actually _is_ output to show?

You can verify by visiting `/api/v2/project_updates/:id`

There's also information sometimes available, separate from the output or tracebacks, on the details panel next to the status.

**Comment by hjkatz** [comment]
Oh! There is something at `/api/v2/project_updates/3341/stdout/` here it is:

```
Identity added: /tmp/awx_3341_7b88dg2g/artifacts/3341/ssh_key_data (/tmp/awx_3341_7b88dg2g/artifacts/3341/ssh_key_data)
ansible-playbook 2.9.17
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/var/lib/awx/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.6.8 (default, Aug 24 2020, 17:57:11) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
Using /etc/ansible/ansible.cfg as config file
host_list declined parsing /tmp/awx_3341_7b88dg2g/inventory/hosts as it did not pass its verify_file() method
script declined parsing /tmp/awx_3341_7b88dg2g/inventory/hosts as it did not pass its verify_file() method
auto declined parsing /tmp/awx_3341_7b88dg2g/inventory/hosts as it did not pass its verify_file() method
Parsed /tmp/awx_3341_7b88dg2g/inventory/hosts inventory source with ini plugin
Skipping callback 'actionable', as we already have a stdout callback.
Skipping callback 'awx_display', as we already have a stdout callback.
Skipping callback 'counter_enabled', as we already have a stdout callback.
Skipping callback 'debug', as we already have a stdout callback.
Skipping callback 'dense', as we already have a stdout callback.
Skipping callback 'dense', as we already have a stdout callback.
Skipping callback 'full_skip', as we already have a stdout callback.
Skipping callback 'json', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'null', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.
Skipping callback 'selective', as we already have a stdout callback.
Skipping callback 'skippy', as we already have a stdout callback.
Skipping callback 'stderr', as we already have a stdout callback.
Skipping callback 'unixy', as we already have a stdout callback.
Skipping callback 'yaml', as we already have a stdout callback.

PLAYBOOK: project_update.yml ***************************************************
2 plays in project_update.yml

PLAY [Update source tree if necessary] *****************************************
META: ran handlers

TASK [update project using git] ************************************************
task path: /tmp/awx_3341_7b88dg2g/project/project_update.yml:34
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: awx
<127.0.0.1> EXEC /bin/sh -c 'echo ~awx && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /var/lib/awx/.ansible/tmp `"&& mkdir "` echo /var/lib/awx/.ansible/tmp/ansible-tmp-1613772303.8304942-12-9529827144520 `" && echo ansible-tmp-1613772303.8304942-12-9529827144520="` echo /var/lib/awx/.ansible/tmp/ansible-tmp-1613772303.8304942-12-9529827144520 `" ) && sleep 0'
Using module file /usr/lib/python3.6/site-packages/ansible/modules/source_control/git.py
<127.0.0.1> PUT /var/lib/awx/.ansible/tmp/ansible-local-3l1sfmdhk/tmp_lkv9ysx TO /var/lib/awx/.ansible/tmp/ansible-tmp-1613772303.8304942-12-9529827144520/AnsiballZ_git.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /var/lib/awx/.ansible/tmp/ansible-tmp-1613772303.8304942-12-9529827144520/ /var/lib/awx/.ansible/tmp/ansible-tmp-1613772303.8304942-12-9529827144520/AnsiballZ_git.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3.6 /var/lib/awx/.ansible/tmp/ansible-tmp-1613772303.8304942-12-9529827144520/AnsiballZ_git.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /var/lib/awx/.ansible/tmp/ansible-tmp-1613772303.8304942-12-9529827144520/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "cmd": [
        "/usr/bin/git",
        "fetch",
        "--tags",
        "origin",
        "refs/heads/*:refs/remotes/origin/*"
    ],
    "invocation": {
        "module_args": {
            "accept_hostkey": true,
            "archive": null,
            "bare": false,
            "clone": true,
            "depth": null,
            "dest": "/var/lib/awx/projects/_6__shared_ansible_repo",
            "executable": null,
            "force": true,
            "gpg_whitelist": [],
            "key_file": null,
            "recursive": true,
            "reference": null,
            "refspec": "refs/heads/*:refs/remotes/origin/*",
            "remote": "origin",
            "repo": "REDACTED",
            "separate_git_dir": null,
            "ssh_opts": null,
            "track_submodules": false,
            "umask": null,
            "update": true,
            "verify_commit": false,
            "version": "hkatz/delete-vm-awx"
        }
    },
    "msg": "Failed to download remote objects and refs:  Warning: Permanently added '[REDACTED]:7999,[REDACTED]:7999' (RSA) to the list of known hosts.\\r\\nFrom ssh://$encrypted$:$encrypted$/systems/ansible\\n ! [rejected]              hkatz/delete-vm-awx -> origin/hkatz/delete-vm-awx  (non-fast-forward)\\n"
}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

I dug into this in the past when originally setting up AWX and this problem actually exists for our main project too: When the project has a rebase on the tip of a branch such as `master` or some branch name, then AWX's playbook to update the repo doesn't work at all due to the `(non-fast-forward)` problem. A fix would be to catch this and support `git pull --rebase` or even a `git checkout --force origin/<branch>` and then `git reset --hard origin/<branch>`.

Currently this is preventing us from being able to update our Project and instead we have to do a full delete + re-clone every update (which is a lot btw...). Additionally, these errors and stdout should be included in the output of failed jobs or otherwise have a prominent place in the Job UI page.

Here's the playbook: https://github.com/ansible/awx/blob/devel/awx/playbooks/project_update.yml#L35

And in the `git` ansible module I think this is a known problem.... so I'm not sure what to do. But it's definitely not great UX behaviour.

**Comment by jakemcdermott** [comment]
Thanks @hjkatz! We'll take a look.

**Comment by AlanCoding** [comment]
@hjkatz If you're still able, could you test if #10005 fixes the problem for you? Thanks.

**Comment by tiagodread** [comment]
I tried to reproduce this problem again without success the project sync is being completed with custom SCM branch and force push
![image](https://user-images.githubusercontent.com/12464858/118876045-2e4f1400-b8c3-11eb-9689-e40bd5244643.png)

**Comment by dkoper** [comment]
FWIW, setting refspec on the project to "+refs/heads/*:refs/remotes/origin/*" (i.e. prefix the default value with a plus, which means override) resolved this error for me on AWX 14.1.0.

**Comment by phemmer** [comment]
Still experiencing this issue on 19.2.2 commit 41321d8.
Overriding the refspec also addressed the issue for me. However note that there should be some `*`s in there. The markdown in @dkoper's comment ate them. The refspec should be: `+refs/heads/*:refs/remotes/origin/*`.

It might also be worth mentioning that this is the default refspec git uses. I don't know why AWX is using something different.
